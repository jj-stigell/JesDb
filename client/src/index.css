@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .link {
    @apply underline mx-1 text-blue-600 hover:text-blue-800 dark:text-blue-300 dark:hover:text-blue-600;
  }
  .title {
    @apply my-2 text-xl text-black font-extrabold dark:text-white;
  }
}

body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.keyword {
  color: red;
}
