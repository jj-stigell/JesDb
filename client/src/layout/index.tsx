import React from 'react'
import { Outlet } from 'react-router-dom'

import Modal from './Modal'
import Menu from './Menu'
import Footer from './Footer'

export default function MainLayout (): React.JSX.Element {
  return (
    <>
      <div
        className="dark:bg-jesdark-900"
        style={{ minHeight: '100vh', display: 'flex', flexDirection: 'column' }}
      >
        <Modal />
        <Menu />
        <Outlet/>
      </div>
      <Footer />
    </>
  )
}
