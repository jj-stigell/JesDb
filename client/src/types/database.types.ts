// DO NOT EDIT THIS FILE
// TYPES ARE AUTOGENERATED FROM DATABASE SCHEMA

export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json | undefined }
  | Json[]

export interface Database {
  public: {
    Tables: {
      notes: {
        Row: {
          audio: string | null
          created_at: string
          external_id: number
          id: number
          image: string | null
          sentence: string
          sentence_furigana: string
          source: Database['public']['Enums']['media_source'] | null
          source_id: string | null
          start_time: number | null
          translation: string | null
        }
        Insert: {
          audio?: string | null
          created_at?: string
          external_id: number
          id?: number
          image?: string | null
          sentence: string
          sentence_furigana: string
          source?: Database['public']['Enums']['media_source'] | null
          source_id?: string | null
          start_time?: number | null
          translation?: string | null
        }
        Update: {
          audio?: string | null
          created_at?: string
          external_id?: number
          id?: number
          image?: string | null
          sentence?: string
          sentence_furigana?: string
          source?: Database['public']['Enums']['media_source'] | null
          source_id?: string | null
          start_time?: number | null
          translation?: string | null
        }
        Relationships: []
      }
      profiles: {
        Row: {
          action_type: Database['public']['Enums']['action_type']
          audio_field: string | null
          created_at: string
          deck_name: string
          expression_field: string | null
          id: number
          image_field: string | null
          name: string
          note_type: string
          reading_field: string | null
          translation_field: string | null
          user_id: string
        }
        Insert: {
          action_type: Database['public']['Enums']['action_type']
          audio_field?: string | null
          created_at?: string
          deck_name: string
          expression_field?: string | null
          id?: number
          image_field?: string | null
          name: string
          note_type: string
          reading_field?: string | null
          translation_field?: string | null
          user_id: string
        }
        Update: {
          action_type?: Database['public']['Enums']['action_type']
          audio_field?: string | null
          created_at?: string
          deck_name?: string
          expression_field?: string | null
          id?: number
          image_field?: string | null
          name?: string
          note_type?: string
          reading_field?: string | null
          translation_field?: string | null
          user_id?: string
        }
        Relationships: [
          {
            foreignKeyName: 'public_profile_user_id_fkey'
            columns: ['user_id']
            isOneToOne: false
            referencedRelation: 'users'
            referencedColumns: ['id']
          },
        ]
      }
      synonyms: {
        Row: {
          synonyms: string[] | null
          term: string
        }
        Insert: {
          synonyms?: string[] | null
          term: string
        }
        Update: {
          synonyms?: string[] | null
          term?: string
        }
        Relationships: []
      }
    }
    Views: {
      [_ in never]: never
    }
    Functions: {
      delete_user: {
        Args: Record<PropertyKey, never>
        Returns: undefined
      }
      search_sentences: {
        Args: {
          keyword: string
        }
        Returns: Array<{
          id: number
          sentence: string
          len: number
          sentence_furigana: string
          translation: string
          audio: string
          image: string
          start_time: number
          source_id: string
          highlighted_sentence: string
          highlighted_language: string
          sentence_furigana_highlight: string
        }>
      }
    }
    Enums: {
      action_type: 'addNote' | 'updateNote'
      media_source: 'youtube'
    }
    CompositeTypes: {
      [_ in never]: never
    }
  }
}

type PublicSchema = Database[Extract<keyof Database, 'public'>]

export type Tables<
  PublicTableNameOrOptions extends
  | keyof (PublicSchema['Tables'] & PublicSchema['Views'])
  | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof (Database[PublicTableNameOrOptions['schema']]['Tables'] &
    Database[PublicTableNameOrOptions['schema']]['Views'])
    : never = never,
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? (Database[PublicTableNameOrOptions['schema']]['Tables'] &
    Database[PublicTableNameOrOptions['schema']]['Views'])[TableName] extends {
      Row: infer R
    }
      ? R
      : never
  : PublicTableNameOrOptions extends keyof (PublicSchema['Tables'] &
  PublicSchema['Views'])
    ? (PublicSchema['Tables'] &
      PublicSchema['Views'])[PublicTableNameOrOptions] extends {
        Row: infer R
      }
        ? R
        : never
    : never

export type TablesInsert<
  PublicTableNameOrOptions extends
  | keyof PublicSchema['Tables']
  | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicTableNameOrOptions['schema']]['Tables']
    : never = never,
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? Database[PublicTableNameOrOptions['schema']]['Tables'][TableName] extends {
    Insert: infer I
  }
    ? I
    : never
  : PublicTableNameOrOptions extends keyof PublicSchema['Tables']
    ? PublicSchema['Tables'][PublicTableNameOrOptions] extends {
      Insert: infer I
    }
      ? I
      : never
    : never

export type TablesUpdate<
  PublicTableNameOrOptions extends
  | keyof PublicSchema['Tables']
  | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicTableNameOrOptions['schema']]['Tables']
    : never = never,
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? Database[PublicTableNameOrOptions['schema']]['Tables'][TableName] extends {
    Update: infer U
  }
    ? U
    : never
  : PublicTableNameOrOptions extends keyof PublicSchema['Tables']
    ? PublicSchema['Tables'][PublicTableNameOrOptions] extends {
      Update: infer U
    }
      ? U
      : never
    : never

export type Enums<
  PublicEnumNameOrOptions extends
  | keyof PublicSchema['Enums']
  | { schema: keyof Database },
  EnumName extends PublicEnumNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicEnumNameOrOptions['schema']]['Enums']
    : never = never,
> = PublicEnumNameOrOptions extends { schema: keyof Database }
  ? Database[PublicEnumNameOrOptions['schema']]['Enums'][EnumName]
  : PublicEnumNameOrOptions extends keyof PublicSchema['Enums']
    ? PublicSchema['Enums'][PublicEnumNameOrOptions]
    : never
