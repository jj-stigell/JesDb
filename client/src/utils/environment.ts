/**
 * Current Node environment.
 * This variable indicates the environment in which the Node application is running.
 * Possible values: 'development', 'production', and 'test'.
 * Defaults to 'development' if the NODE_ENV environment variable is not set.
 */
export const NODE_ENV: string = process.env.REACT_APP_NODE_ENV ?? 'development'

/**
 * Supabase URL. This is the URL to your Supabase project.
 * You can find this key in your Supabase project settings.
 */
export const SUPABASE_URL: string = process.env.REACT_APP_SUPABASE_URL ?? ''

/**
 * Supabase API key. Only use public key here. Do not expose your private key.
 * This key is used to authenticate with the Supabase API.
 * You can find this key in your Supabase project settings.
 */
export const SUPABASE_KEY: string = process.env.REACT_APP_SUPABASE_KEY ?? ''

/**
 * Auth callback URL. This is the URL that the user will be redirected to after
 * they authenticate with Supabase.
 * This URL must be added to the list of "Redirect URIs" in your Supabase project settings.
 */
export const AUTH_CALLBACK_URL: string =
  process.env.REACT_APP_AUTH_CALLBACK_URL ?? 'http://localhost:3000/auth/callback'
