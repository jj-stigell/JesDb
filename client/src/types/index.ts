import { Database } from './database.types'

/**
 * No params for deckNames, modelNamesAndIds, modelNames
 */
export type Action =
  'deckNames' |
  'updateNote' |
  'modelNamesAndIds' |
  'modelFieldNames' |
  'modelNames' |
  'findNotes' |
  'addNote' |
  'getNoteTags'

export interface UpdateNoteParams {
  note: {
    id: number
    fields: Record<string, string>
    tags: string[]
  }
}

export interface AddNoteParams {
  note: {
    deckName: string
    modelName: string
    fields: Record<string, string>
    options?: {
      allowDuplicate?: boolean
      duplicateScope?: string
      duplicateScopeOptions?: {
        deckName?: string
        checkChildren?: boolean
        checkAllModels?: boolean
      }
    }
    tags?: string[]
    audio?: Array<{
      url: string
      filename: string
      skipHash: string
      fields: string[]
    }>
    video?: Array<{
      url: string
      filename: string
      skipHash: string
      fields: string[]
    }>
    picture?: Array<{
      url: string
      filename: string
      skipHash: string
      fields: string[]
    }>
  }
}

export type NoteParams = UpdateNoteParams | AddNoteParams

interface ModelFieldNamesParams {
  modelName: string
}

interface FindNotesParams {
  query: string
}

interface GetNoteTagsParams {
  note: number
}

interface ActionParamsMap {
  deckNames: undefined
  updateNote: UpdateNoteParams
  modelNamesAndIds: undefined
  modelFieldNames: ModelFieldNamesParams
  modelNames: undefined
  findNotes: FindNotesParams
  addNote: AddNoteParams
  getNoteTags: GetNoteTagsParams
}

export interface ErrorResult {
  result: null
  error: string
}

interface ResultStringArray {
  result: string[]
  error: null
}

export interface FindNotes {
  result: number[]
  error: null
}

export interface AddNote {
  result: number
  error: null
}

export interface UpdateNoteResult {
  result: null
  error: null
}

interface ModelNamesAndIdsResult {
  result: Array<Record<string, number>>
  error: null
}

interface ActionResultMap {
  deckNames: ResultStringArray
  updateNote: UpdateNoteResult
  modelNamesAndIds: ModelNamesAndIdsResult
  modelFieldNames: ResultStringArray
  modelNames: ResultStringArray
  findNotes: FindNotes
  addNote: AddNote
  getNoteTags: ResultStringArray
}

export type ExecuteActionParams<T extends Action> =
  T extends keyof ActionParamsMap ? ActionParamsMap[T] : never

export type ExecuteActionReturnType<T extends Action> = ActionResultMap[T]

export interface Body<T extends Action> {
  action: Action
  version: number
  params?: ExecuteActionParams<T>
}

// Represents the note type that search function returns
export type Note = Database['public']['Functions']['search_sentences']['Returns'][0]

export type Profile = Database['public']['Tables']['profiles']['Row']

export type ActionType = Database['public']['Enums']['action_type']

export interface NewProfile extends Omit<Profile, 'id' | 'user_id' | 'created_at'> {}

export interface ExportResult {
  result: string | null
  error: string | null
}

export interface Filters {
  sort: 'relevance' | 'shortness' | 'longness'
  onlyTranslated: boolean
}
