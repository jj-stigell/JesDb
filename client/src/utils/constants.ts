import { Note, Profile } from 'src/types'

export const repoLink = 'https://github.com/jj-stigell/JesDb'

export const feedbackLink = 'https://forms.gle/SVNcSrujPHoy6UuG9'

export const ankiConnectDefaultAddress = 'http://127.0.0.1:8765'

// This one has highlighted_language set to 'ja' therefore it will have sentence_furigana_highlight
export const exampleNoteJa: Note = {
  id: 1863,
  sentence: '昔から 自動車の会社をやっていて',
  len: 16,
  sentence_furigana: '昔[むかし]から 自動車[じどうしゃ]の 会社[かいしゃ]をやっていて',
  translation: 'long time a go I had a car company',
  audio: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/gBumdOWWMhY/gBumdOWWMhY_s102_e104.mp3`,
  image: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/gBumdOWWMhY/gBumdOWWMhY_s102_e104.jpg`,
  start_time: 102,
  source_id: 'gBumdOWWMhY',
  highlighted_sentence: '昔から <span class="keyword">自動車</span>の会社をやっていて',
  highlighted_language: 'ja',
  sentence_furigana_highlight: '昔[むかし]から <span class="keyword">自動車</span>[じどうしゃ]の 会社[かいしゃ]をやっていて'
}

// This one has highlighted_language set to 'en' therefore sentence_furigana_highlight is empty
export const exampleNoteEn: Note = {
  id: 1135,
  sentence: 'それが怖くなったのでございます',
  len: 15,
  sentence_furigana: 'それが 怖[こわ]くなったのでございます',
  translation: 'I was scared of what was waiting for me.',
  audio: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/ylEimFz8KOo/ylEimFz8KOo_s159_e162.mp3`,
  image: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/ylEimFz8KOo/ylEimFz8KOo_s159_e162.jpg`,
  start_time: 159,
  source_id: 'ylEimFz8KOo',
  highlighted_sentence: 'I was <span class="keyword">scared</span> of what was waiting for me.',
  highlighted_language: 'en',
  sentence_furigana_highlight: ''
}

export const exampleProfile: Profile = {
  id: 1,
  name: 'Test Profile',
  deck_name: 'Test Deck',
  note_type: 'Basic',
  action_type: 'addNote',
  expression_field: 'expression',
  reading_field: 'reading',
  translation_field: 'translation',
  audio_field: 'audio',
  image_field: 'image',
  created_at: '2021-09-01T00:00:00.000Z',
  user_id: '1af8625c-0771-4c53-9851-42274e8e2385'
}
