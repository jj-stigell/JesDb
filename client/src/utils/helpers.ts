import { NoteParams, Note, Profile } from 'src/types'

export function getFileName (url: string): string {
  const regex = /[^/]+\.(mp3|jpg|png)$/
  const match = url.match(regex)

  if (match !== null) {
    return match[0]
  } else {
    return `${(+new Date()).toString(36)}.${url.split('.').pop() ?? 'file'}`
  }
}

export function scrollToTop (): void {
  window.scrollTo({
    top: 0,
    behavior: 'smooth'
  })
}

export function constructParams (
  example: Note,
  profile: Profile
): NoteParams {
  const fields: any = {}
  const audio: any = {}
  const picture: any = {}

  if (profile.expression_field !== 'omit' && profile.expression_field !== null) {
    fields[profile.expression_field] = example.sentence
  }

  if (profile.reading_field !== 'omit' && profile.reading_field !== null) {
    fields[profile.reading_field] = example.sentence_furigana
  }

  if (profile.translation_field !== 'omit' && profile.translation_field !== null) {
    fields[profile.translation_field] = example.translation
  }

  if (profile.audio_field !== 'omit' && example.audio !== null && profile.audio_field !== null) {
    fields[profile.audio_field] = ''
    audio.url = example.audio
    audio.filename = getFileName(example.audio)
    audio.fields = [profile.audio_field]
  }

  if (profile.image_field !== 'omit' && example.image !== null && profile.image_field !== null) {
    fields[profile.image_field] = ''
    picture.url = example.image
    picture.filename = getFileName(example.image)
    picture.fields = [profile.image_field]
  }

  const params: any = {
    note: {
      fields,
      tags: ['JesDb', example.source_id],
      audio: [audio],
      picture: [picture]
    }
  }

  return params
}

export function parseRuby (input: string): string {
  const regex = /(^|\s|<[^>]+>)([^[<\s]+)\[([^\]]+)\]/g
  return input.replace(regex, (match, p1, kanji: string, furigana: string) => {
    const prefix: string = p1.trim().length > 0 ? p1 : ''
    return `${prefix}<ruby>${kanji}<rt>${furigana}</rt></ruby>`
  })
}
