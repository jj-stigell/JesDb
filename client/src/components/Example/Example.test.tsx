import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import Example, { IExample } from './Example'
import { exampleNoteJa, exampleProfile } from '../../utils/constants'
import { Profile } from 'src/types'

const mockProfiles: Profile[] = [
  {
    ...exampleProfile,
    id: 1,
    name: 'Profile 1'
  },
  {
    ...exampleProfile,
    id: 2,
    name: 'Profile 2'
  }
]

const defaultProps: IExample = {
  note: exampleNoteJa,
  profiles: mockProfiles,
  isPlaying: false,
  showCompact: false,
  showFurigana: false,
  exportNote: jest.fn(),
  playAudio: jest.fn()
}

// Mock HTMLMediaElement methods
window.HTMLMediaElement.prototype.play = jest.fn()
window.HTMLMediaElement.prototype.pause = jest.fn()

describe('Example Component', () => {
  it('renders component properly, full view', () => {
    const { asFragment } = render(<Example {...defaultProps} />)
    const furiganaComponent = screen.getByTestId('furigana-component')

    expect(asFragment()).toMatchSnapshot()
    expect(furiganaComponent).toBeInTheDocument()
    expect(furiganaComponent.innerHTML).toContain(exampleNoteJa.highlighted_sentence)
  })

  it('renders component properly, compact view', () => {
    const { asFragment } = render(<Example {...defaultProps} showCompact={true} />)
    const furiganaComponent = screen.getByTestId('furigana-component')

    expect(asFragment()).toMatchSnapshot()
    expect(furiganaComponent).toBeInTheDocument()
    expect(furiganaComponent.innerHTML).toContain(exampleNoteJa.highlighted_sentence)
  })

  it('calls playAudio when play button is clicked', () => {
    render(<Example {...defaultProps} />)
    const playButton = screen.getByRole('button', { name: 'Play audio' })
    fireEvent.click(playButton)

    expect(defaultProps.playAudio).toHaveBeenCalledWith(defaultProps.note.audio)
  })

  it('calls stopAudio when stop button is clicked', () => {
    render(<Example {...defaultProps} isPlaying={true} />)
    const stopButton = screen.getByRole('button', { name: 'Stop audio' })
    fireEvent.click(stopButton)

    expect(defaultProps.playAudio).toHaveBeenCalled()
  })

  it('calls exportNote when export profile is selected', () => {
    render(<Example {...defaultProps} />)
    const exportButton = screen.getByText(/export to anki/i)
    fireEvent.click(exportButton)
    const profileOption = screen.getByText(/Profile 1/i)
    fireEvent.click(profileOption)

    expect(defaultProps.exportNote).toHaveBeenCalledWith(mockProfiles[0], exampleNoteJa)
  })

  it('calls exportNote when other export profile is selected', () => {
    render(<Example {...defaultProps} />)
    const exportButton = screen.getByText(/export to anki/i)
    fireEvent.click(exportButton)
    const profileOption = screen.getByText(/Profile 2/i)
    fireEvent.click(profileOption)

    expect(defaultProps.exportNote).toHaveBeenCalledWith(mockProfiles[0], exampleNoteJa)
  })
})
