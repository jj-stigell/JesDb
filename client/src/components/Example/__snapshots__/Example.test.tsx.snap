// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example Component renders component properly, compact view 1`] = `
<DocumentFragment>
  <div
    class="
        relative flex flex-col md:flex-row md:space-x-1 space-y-1 md:space-y-0
        rounded-xl shadow-lg p-1 mx-auto border-blue-300 border bg-blue-100
        my-3 dark:bg-jesdark-600 dark:border-jesdark-700
      "
  >
    <div
      class="w-full bg-inherit flex flex-col pl-2"
    >
      <div>
        <div
          aria-live="polite"
          class="text-lg dark:text-white"
          data-testid="furigana-component"
          lang="ja"
        >
          昔から 
          <span
            class="keyword"
          >
            自動車
          </span>
          の会社をやっていて
        </div>
      </div>
      <div
        class="text-xs dark:text-white"
      >
        long time a go I had a car company
      </div>
    </div>
    <div
      class="flex flex-row gap-3"
    >
      <div>
        <button
          aria-label="Play audio"
          class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-blue-700 text-white focus:ring-4 focus:ring-blue-300 enabled:hover:bg-blue-800 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 rounded-lg"
          type="button"
        >
          <span
            class="flex items-stretch transition-all duration-200 rounded-md px-2 py-1 text-xs"
          >
            <svg
              class="my-0.5"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80V432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="w-fit"
        data-testid="flowbite-tooltip-target"
      >
        <button
          class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-blue-700 text-white focus:ring-4 focus:ring-blue-300 enabled:hover:bg-blue-800 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 rounded-lg"
          data-testid="source-button"
          type="button"
        >
          <span
            class="flex items-stretch transition-all duration-200 rounded-md px-2 py-1 text-xs"
          >
            <svg
              aria-hidden="true"
              class="my-0.5"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M4 3a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V5a2 2 0 00-2-2H4zm3 2h6v4H7V5zm8 8v2h1v-2h-1zm-2-2H7v4h6v-4zm2 0h1V9h-1v2zm1-4V5h-1v2h1zM5 5v2H4V5h1zm0 4H4v2h1V9zm-1 4h1v2H4v-2z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="absolute z-10 inline-block rounded-lg px-3 py-2 text-sm font-medium shadow-sm transition-opacity duration-300 invisible opacity-0 bg-gray-900 text-white dark:bg-gray-700"
        data-testid="flowbite-tooltip"
        id=":r2:"
        role="tooltip"
        style="position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="relative z-20"
        >
          Open source video in new tab
        </div>
        <div
          class="absolute z-10 h-2 w-2 rotate-45 bg-gray-900 dark:bg-gray-700"
          data-testid="flowbite-tooltip-arrow"
          style="bottom: -4px;"
        >
           
        </div>
      </div>
      <div
        class="hidden sm:block"
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          aria-label="Export to Anki dropdown"
          class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-blue-700 text-white focus:ring-4 focus:ring-blue-300 enabled:hover:bg-blue-800 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 rounded-lg w-fit"
          data-testid="flowbite-dropdown-target"
          id=":r5:"
          type="button"
        >
          <span
            class="flex items-stretch transition-all duration-200 rounded-md px-2 py-1 text-xs"
          >
            <svg
              baseProfile="tiny"
              class="h-4 w-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              version="1.2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 16.5v.5c1.691-2.578 3.6-3.953 6-4v3c0 .551.511 1 1.143 1 .364 0 .675-.158.883-.391 1.933-2.029 5.974-6.109 5.974-6.109s-4.041-4.082-5.975-6.137c-.208-.205-.518-.363-.882-.363-.632 0-1.143.447-1.143 1v3c-4.66 0-6 4.871-6 8.5zM5 21h14c.553 0 1-.448 1-1v-6.046c-.664.676-1.364 1.393-2 2.047v2.999h-12v-12h7v-2h-8c-.553 0-1 .448-1 1v14c0 .552.447 1 1 1z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="ml-2 h-4 w-4"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 9l-7 7-7-7"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Example Component renders component properly, full view 1`] = `
<DocumentFragment>
  <div
    class="
        relative flex flex-col md:flex-row md:space-x-1 space-y-1 md:space-y-0
        rounded-xl shadow-lg p-1 mx-auto border-blue-300 border bg-blue-100
        my-3 dark:bg-jesdark-600 dark:border-jesdark-700
      "
  >
    <div
      class="w-full md:w-1/2 bg-inherit grid place-items-center"
    >
      <div
        class="
        flex items-center justify-center w-full h-44
        bg-gray-300 dark:bg-gray-700 animate-pulse rounded-xl"
      >
        <svg
          aria-hidden="true"
          class="w-10 h-10 text-gray-200 dark:text-gray-600"
          fill="currentColor"
          viewBox="0 0 20 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="
              M18 0H2a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2Zm-5.5
              4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm4.376 10.481A1 1 0 0 1 16 15H4a1 1 0 0
              1-.895-1.447l3.5-7A1 1 0 0 1 7.468 6a.965.965 0 0 1 .9.5l2.775 4.757 1.546-1.887a1
              1 0 0 1 1.618.1l2.541 4a1 1 0 0 1 .028 1.011Z"
          />
        </svg>
      </div>
      <img
        alt="image for example id 1863"
        class="rounded-xl"
        src="https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/gBumdOWWMhY/gBumdOWWMhY_s102_e104.jpg"
      />
    </div>
    <div
      class="w-full md:w-2/3 bg-inherit flex flex-col space-y-2 p-2"
    >
      <div
        class="flex justify-between item-center"
      >
        <p
          class="text-xs dark:text-white"
        >
          Source:
          <a
            class="
                text-blue-600 hover:text-blue-900 underline ml-1
                dark:text-blue-300 dark:hover:text-blue-400
              "
            data-testid="source-link"
            href="https://youtube.com/watch?v=gBumdOWWMhY&t=102s"
            rel="noreferrer"
            target="_blank"
          >
            gBumdOWWMhY
          </a>
        </p>
        <div
          class="hidden sm:block"
        >
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            aria-label="Export to Anki dropdown"
            class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-blue-700 text-white focus:ring-4 focus:ring-blue-300 enabled:hover:bg-blue-800 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 rounded-lg w-fit"
            data-testid="flowbite-dropdown-target"
            id=":r1:"
            type="button"
          >
            <span
              class="flex items-stretch transition-all duration-200 rounded-md px-2 py-1 text-xs"
            >
              Export to Anki
              <svg
                aria-hidden="true"
                class="ml-2 h-4 w-4"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 9l-7 7-7-7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </div>
        <button
          aria-label="Play audio"
          class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-blue-700 text-white focus:ring-4 focus:ring-blue-300 enabled:hover:bg-blue-800 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 rounded-lg"
          type="button"
        >
          <span
            class="flex items-stretch transition-all duration-200 rounded-md px-2 py-1 text-xs"
          >
            <svg
              class="my-0.5"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80V432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="flex justify-between item-center mt-2"
      >
        <div
          aria-live="polite"
          class="text-lg dark:text-white"
          data-testid="furigana-component"
          lang="ja"
        >
          昔から 
          <span
            class="keyword"
          >
            自動車
          </span>
          の会社をやっていて
        </div>
      </div>
      <div
        class="text-xs dark:text-white"
      >
        long time a go I had a car company
      </div>
    </div>
  </div>
</DocumentFragment>
`;
